[Doc Feature] Bot sessions (#14081)
MicrosoftDocs/msteams-docs: msteams-platform/bots/how-to/conversations/bot-session.md

---
title: Use sessions in bot conversations in Microsoft Teams
description: Sessions enable bots in Microsoft Teams to organize one-on-one conversations into separate, focused chats, helping maintain context, improve response relevance, and support multi-task workflows.
ms.author: vikasalmal
ms.localizationpriority: high
ms.topic: conceptual
ms.date: 04/28/2026
---

<!-- markdownlint-disable MD024 -->

# Use sessions in bot conversations in Microsoft Teams

Sessions allow users to organize conversations with your bot into independent discussion chats within a single one-on-one thread. Each session maintains its own conversation history, helping reduce context length, improve response relevance, and provide a structured way for users to manage multiple tasks or workflows with your bot.

> Sessions help bots align with conversation patterns that users already experience in modern AI assistants, where conversations are separated by task or subject instead of grouped into a single continuous thread.

Reviewer: These conversation patterns developed in part because of the way AI assistants work and are implemented (context windows etc.), so now that these patterns exist in Teams, how does the developer's agent implementation need to change to a) simply accommodate the pattern, and b) take best advantage of it, beyond the user experience aspects of keeping tasks together? For example, the SDK docs currently have an article called [Keeping State](https://microsoft.github.io/teams-sdk/typescript/in-depth-guides/ai/keeping-state); how does that play with sessions?

## Agents navigation integration

When topics are enabled, Teams introduces an **Agents** navigation experience that aggregates bot-related activity.

The Agents navigation:

- Displays all topic-related activity for the bot.

- Highlights new activity across conversations.

- Allows users to follow or unfollow topics.

Unfollowing a topic removes it from navigation but doesn't delete the conversation.

## Deep linking to topics

Deep links that previously opened one-on-one bot chats continue to work after topics are enabled. When users open an existing deep link:

- Teams routes the user to the correct topic.

- Existing links remain backward compatible.

- Users maintain access to historical conversations.
-->
## Existing chat migration to default session

When sessions are enabled for an existing bot, previous messages are automatically migrated to a default session. This migration ensures that users don't lose conversation history after upgrading the app.

During migration:

- Teams groups existing messages into a default session
- Users can continue chatting without interruption
- New sessions can be created after migration completes

> ## Best practices

Reviewer: There's a key best practice/decision guide missing here - when should a developer enable sessions for their agent, and why?

| 403 | `BotDisabledByAdmin` | Tenant admin has disabled the bot. |
| 403 | `BotNotInConversationRoster` | Bot is not installed in the conversation. |
| 404 | `ConversationNotFound` | The conversation thread doesn't exist. |
| 429 | `Throttled` | Rate limit exceeded. Retry after the indicated delay. |
| 500 | `ServiceError` | Unexpected server error. |
| 502 | — | Upstream service timed out while verifying bot session support. |

## Best practices

### Organize sessions around user tasks

Create sessions that represent clear goals or workflows. For example, you might create separate sessions for individual tickets, pull requests, or customer requests.

Clear session separation helps users quickly locate and resume conversations.

### Limit creating too many sessions

Create a new session only when it represents a distinct task or workflow that benefits from its own conversation context. Avoid creating sessions for transient updates or single-message interactions. Too many sessions make it harder for users to find relevant conversations in the sessions panel.

### Use a dedicated session for notifications

> If your bot sends updates to users based on activity outside of Teams and unrelated to chat history, such as CI/CD results, service alerts, or external system notifications, route them to a single dedicated notifications session instead of creating a new session for each update. Creating a session per notification floods the sessions panel and makes it difficult for users to find task-specific conversations. Store the conversation ID of the notifications session after you create it, and reuse that ID for subsequent messages.

Reviewer: Combine this with the one above, they're both about the decision process for proactive

---

> title: Use sessions in bot conversations in Microsoft Teams

Reviewer: Given that this is a brand new article, let's take the opportunity to use _agent_ everywhere instead of _bot_.

Author: Agreed on using 'agent' in prose for this new article. A few items will need to retain 'bot' due to the underlying SDK and manifest schema (botId, "bots": [...], API error codes). I'll update all user-facing prose to use 'agent' while preserving technical identifiers. Should the file also be renamed (e.g., agent-sessions.md), or keep [bot-session.md] for now?

Reviewer: >A few items will need to retain 'bot' due to the underlying SDK and manifest schema (botId, "bots": [...], API error codes).

Acknowledged, and agree.

>Should the file also be renamed (e.g., agent-sessions.md), or keep [bot-session.md] for now?

Please rename.

description: Sessions enable bots in Microsoft Teams to organize one-on-one conversations into separate, focused chats, helping maintain context, improve response relevance, and support multi-task workflows.
ms.author: vikasalmal
ms.localizationpriority: high
ms.topic: article
ms.date: 04/28/2026
zone_pivot_groups: teams-sdk-languages
---

<!-- markdownlint-disable MD024 -->

> # Use sessions in bot conversations in Microsoft Teams

Reviewer: ```suggestion
# Manage multiple user conversations with sessions
```

Sessions allow users to organize conversations with your bot into independent discussion chats within a single one-on-one chat. Each session maintains its own conversation history, helping reduce context length, improve response relevance, and provide a structured way for users to manage multiple tasks or workflows with your bot.

> Sessions help bots align with conversation patterns that users already experience in modern AI assistants, where conversations are separated by task or subject instead of grouped into a single continuous chat.

Reviewer: ```suggestion
By default, one-on-one chat between a user and an agent takes place in a single long-running conversation. With sessions, an agent can hold multiple independent one-on-one conversations with a user.

Sessions offer users a structured way to manage multiple tasks or workflows with an agent, similar to other modern AI assistant experiences. Encouraging users to organize their interactions into shorter, more focused contexts can also improve the quality of LLM-generated responses.

Sessions is an optional feature that must be enabled for an agent in its manifest configuration.
```

In this article:

- [Enable sessions for your bot](#enable-sessions-for-your-bot)
- [Send and receive messages in sessions](#send-and-receive-messages-in-sessions)
- [Create sessions proactively](#create-sessions-proactively)
- [Session control](#session-control)
- [Detect session support through install events](#detect-session-support-through-install-events)
- [Error codes](#error-codes)
- [Best practices](#best-practices)

> 

Reviewer: ```suggestion
```

We really don't need these on Learn, this is what the sidebar is for.

-->
## Enable sessions for your bot

Sessions are an opt-in capability that you enable through your app manifest.  
After you enable sessions and publish the updated app, users see the sessions experience after they install or upgrade the app.

> [!IMPORTANT]
> After your bot opts in to sessions, we recommend keeping the feature enabled. Once sessions are enabled on a chat, there is no way to revert it to the exact state of a regular 1:1 chat. Opting out of sessions is **not supported at GA**. Opt-out support is planned as a fast follow-up after GA.

Update your app manifest and set the `supportsSessions` property to `true`.

```json
{
  "bots": [
    {
      "botId": "{{BOT_ID}}",
      "supportsSessions": true
    }
  ]
}
```

After you update the manifest, package and republish your app through the [Developer Portal for Teams](https://dev.teams.microsoft.com/) or [Teams admin center](https://admin.teams.microsoft.com/). Users see the sessions experience after they install or upgrade the app.

> [!IMPORTANT]
>
> Bots that don't enable sessions continue to use the legacy single chat experience.

> [!NOTE]

> > When sessions are enabled for an existing bot, Teams automatically converts the existing chat history into a default session. No action is required and users don't lose any conversation history.

Reviewer: Just put these in regular text; studies show that most uses of these alerts has the opposite effect from what's desired.

Author: The IMPORTANT/NOTE alert formatting here was specifically requested by the engineering team during an earlier review round, and the intent was to make the 'no opt-out' warning highly visible to developers. How do you want me to act on this now?

Reviewer: I'm OK with the no-op-out warning being in an alert, you could honestly even upgrade it to a warning, but this one doesn't call for it; we already make it clear that sessions are opt in and optional.

You can loop me into any PM/eng conversations as needed and I will negotiate on this kind of thing.

> ## User experience after enabling sessions

Reviewer: Move this to be the first section after the intro.
```suggestion
## User experience
```

> After sessions are enabled, users see new controls in the chat header that allow them to create and switch between sessions.

Reviewer: ```suggestion
When chatting one-on-one with an agent that supports sessions, controls in the header enable users to create and switch between sessions.
```

- **Switch between sessions**: Move between active sessions through the sessions panel without losing context.
- **Track unread activity**: Sessions with new messages are highlighted and badged so users can identify pending updates.
- **Navigate to specific activity**: Select a notification or session entry to jump directly to the relevant message.
- **Access from personal app**: Open the bot through the personal app entry point. Sessions are fully supported in the personal app experience, including deep links.
- **Search across sessions**: Find messages across all sessions using Teams search.

All existing one-on-one bot chat capabilities continue to work within sessions.

The first message in a session becomes the session title, and the latest message appears as the preview in the sessions panel. Sessions created proactively by the bot appear as unread for the user.

> [!NOTE]
> Sessions are distinct from threaded replies in channels. A session is a full, independent conversation context within a 1:1 chat and not a reply chain under a single message.

## Send and receive messages in sessions

> When sessions are enabled, incoming activities include a session-scoped conversation ID. The conversation ID represents the current session and must be treated as an opaque value. Your bot shouldn't parse or construct the conversation ID manually.

Reviewer: The first sentence here should more or less say that session participation is effectively automatic and requires no special implementation support.

Use the conversation ID from the incoming activity when responding.

::: zone pivot="teams-sdk-csharp"

```csharp
var conversationId = context.Activity.Conversation.Id;
await app.Send(conversationId, "Hello from Bot");
```

::: zone-end

::: zone pivot="teams-sdk-typescript"

```typescript
const conversationId = activity.conversation.id;
await app.send(conversationId, "Hello from Bot");
```

::: zone-end

::: zone pivot="teams-sdk-python"

```python
conversation_id = ctx.activity.conversation.id
await app.send(conversation_id, "Hello from Bot")
```

::: zone-end

Responding inside a session works the same way as responding in a normal one-on-one chat. When your bot replies using the conversation ID from the incoming activity, Teams automatically delivers the message to the correct session. Each session maintains independent conversation context.

> ### How conversationId changes with sessions

Reviewer: This is too much detail for something we just told developers to treat as opaque, and repeats a few of the points too many times.  I suggest getting rid of both of the H3s and condensing this way down.

Author: Agreed this is too verbose for an opaque value. I'll remove both H3 headings and condense to a brief paragraph covering the two key developer-facing behaviors: (1) each session = unique conversationId, no separate sessionId; (2) previously cached IDs still work, routing to the default session. The detail was provided by the engineering team during review. Do you want me to condense it down?

Reviewer: Yes, this sounds good.

>no separate sessionId

I wouldn't use this in the explanation - I totally understand what this kind of explanation is meant to do, but referencing a property that doesn't actually exist (even just to make a point) can confuse people.

**Without sessions**, all messages in a 1:1 chat share a single, static `conversationId`.

**With sessions**, each session gets its own unique `conversationId`. When the bot receives any activity, `activity.conversation.id` tells it which session the activity belongs to.

There is no separate `sessionId` field. The session identity is encoded within the `conversationId`. Each session has a unique `conversationId`, and that is the session's identifier. Treat each unique `conversationId` as a unique session. No parsing is needed. Use the opaque string as-is for storage and routing.

### Using a cached conversationId after sessions are enabled

If your bot cached a `conversationId` from before the user opted into sessions, that cached ID still works. Messages sent to it land in the **default session**, which is the very first session created when the 1:1 chat becomes sessions-enabled. The default session acts as a catch-all that contains all pre-existing messages.

This means bots don't break when sessions are enabled. Existing cached conversationIds continue to function, routing messages to the default session.

> ### HTTP API reference

Reviewer: This commands a lot of vertical space; is it needed if all the support for this is in the opaque conversation ID? Can we just defer to the general messaging documentation?

Author: The 'Create a new session' HTTP snippet is sessions-specific. The send/reply snippets show the standard pattern with a session conversationId. The get-members call is the only place this HTTP endpoint is documented in our bot docs. I can trim the section to just the create and get-members calls, or keep all four but reduce vertical space by removing the repeated headers/auth lines. Which would you prefer? 

Engineering team provided these and they're the only HTTP-level reference for these operations.

Reviewer: Understood, thanks. Leave it in for now and maybe we'll revisit.

    "type": "message",
    "text": "Starting a new session"
  }
}
```

**Response (201 Created)**:

```json
{
  "id": "<session-conversationId>",
  "activityId": "1749012345678"
}
```

The `id` is the session's conversationId, an opaque, encrypted string. Store it and use it for all subsequent operations within this session.

#### Send a message within an existing session

```http
POST {serviceUrl}/v3/conversations/{sessionConversationId}/activities
Authorization: Bearer {bot-token}
Content-Type: application/json

{
  "type": "message",
  "text": "Follow-up message in this session"
}
```

#### Reply to a specific activity

```http
POST {serviceUrl}/v3/conversations/{sessionConversationId}/activities
Authorization: Bearer {bot-token}
Content-Type: application/json

{
  "type": "message",
  "text": "Reply to your question",
  "replyToId": "1749012345678"
}
```

#### Get members

```http
GET {serviceUrl}/v3/conversations/{sessionConversationId}/members
Authorization: Bearer {bot-token}
```

> [!NOTE]
> Member operations (`GetMembers`, `GetMember`, `GetProfile`) work the same regardless of which session conversationId you use. They always resolve to the underlying 1:1 chat.

> ## Create sessions proactively

Reviewer: This should lead by stating the fundamental concept that the existing _create conversation_ operation, applied to 1:1 conversations, creates new sessions.

I just noticed that "Create the conversation" in the Proactive messages article, linked from further down, doesn't actually document how to create a conversation. Should some of this information be moved there?

Overall this section proactive section needs a bit of reorganization, it states a lot of things multiple times and has a few important concepts buried in the middle of the note block.

Author: Agreed on all three points. I'll:

- Reorder to lead with 'the existing create-conversation operation creates a new session'

- Clean up the repetition in the note block

Regarding moving content to the Proactive messages article, that article currently doesn't show how to create a conversation with sessions. I can either move the code samples there (and reference from here) or keep them here and maintain the cross-reference link we already added. Which would you prefer?

Reviewer: >I can either move the code samples there (and reference from here) or keep them here and maintain the cross-reference link we already added. Which would you prefer?

Go ahead and do the latter, I don't want to introduce a bunch more reorganization work. Can revisit later if needed.

Sessions are an opt-in feature. Two conditions must be met before a session can be created:

> 1. **The bot must be installed for the user.** If the bot is not installed, session creation is skipped and a regular 1:1 conversation is created instead (no error returned).

Reviewer: > and a regular 1:1 conversation is created instead

What exactly does this mean for a user who doesn't even have the bot installed? Do they get the message?

Author: Will check with the feature team. The original guidance was that calling CreateConversation 'silently falls back to a regular 1:1 conversation' when the bot isn't installed. But you're right that this needs clarification: can a bot message a user who hasn't installed it at all? Or does this fallback only apply when the bot is installed but sessions haven't activated for that user yet? 

@yogapat, could you please help us on this?

2. **The bot's manifest must declare `supportsSessions: true`.**

> Both conditions are checked server-side. A bot that declares `supportsSessions: true` but is not installed for the target user will silently fall back to a regular 1:1 conversation. If both conditions are met, calling the create conversation API with one message activity creates a new session.

Reviewer: Isn't it critical that it have only one _member_?

```

::: zone-end

::: zone pivot="teams-sdk-typescript"

```typescript
// Including exactly one initial `activity` is required for creating a session
// It becomes the first message of the new session.
const resource = await app.api.conversations.create({
  members: [{ id: userId, role: 'user', name: '' }],
  tenantId,
  activity: { type: 'message', text: 'Hello! Starting a new session.' },
});

// resource.id is the session conversationId
const sessionConversationId = resource.id;

await app.send(sessionConversationId, 'This message is part of the session.');
```

::: zone-end

::: zone pivot="teams-sdk-python"

[TODO]

::: zone-end

> [!NOTE]
>
> The Teams SDK calls the same `POST /v3/conversations` endpoint as regular 1:1 creation. No special flag is needed. The server creates a session automatically when the bot is session-enabled and installed.
> The `activity` field is **required** to have exactly one message activity, which becomes the first message of the new session.
> The returned id (`resource.Id` in C#, `resource.id` in TypeScript) is the session-specific conversationId. Use it for all follow-up messages in that session (via `app.Send` / `app.send`).

> ## Session control

Reviewer: I would remove this, it's just restating everything from above.

Author: There was a question from the engineering "Can bots control session reuse vs new session?" The table format was a nice quick reference. If you want, I can remove this as it is just redundant of information.

Reviewer: Yes, please remove.

>"Can bots control session reuse vs new session?"

This is a great basic question about the overall capability; the table isn't the right approach to it but they are right that it should be clearly addressed. Let's do this:

- Revise the intro sentence "Sessions offer users a structured way to manage multiple tasks or workflows with an agent" to "Sessions offer users and agents a structured way to manage multiple tasks or workflows"
- Add a sentence in _user experience_ after the screenshots that simply states that agents can also proactively create new sessions with the user.

The topics panel provides an overview of all active topics. When new messages or topics are created:

- The topics panel icon displays a notification badge.

- Unread topics appear in bold.

- Users can open topics directly from the panel.

### Activity feed integration

Topic activity, such as mentions or reactions, appears in the Teams activity feed. Selecting an activity notification opens the relevant topic and message, ensuring users can quickly resume conversations.

## Agents navigation integration

When topics are enabled, Teams introduces an **Agents** navigation experience that aggregates bot-related activity.

The Agents navigation:

- Displays all topic-related activity for the bot.

- Highlights new activity across conversations.

- Allows users to follow or unfollow topics.

Unfollowing a topic removes it from navigation but doesn't delete the conversation.

## Deep linking to topics

Deep links that previously opened one-on-one bot chats continue to work after topics are enabled. When users open an existing deep link:

- Teams routes the user to the correct topic.

- Existing links remain backward compatible.

- Users maintain access to historical conversations.
-->

> ## Detect session support through install events

Reviewer: All of this general content about channelData.app.version seems like it should be documented somewhere else, and this would very briefly refer to it and say that it's the mechanism for detection of this specific feature.

Author: Agreed, channelData.app.version is general platform info that should live in a standalone reference. However, it's not documented anywhere else in the repo currently (the engineering team provided it during this review). For this PR, I can slim this section down to a brief paragraph stating that channelData.app.version is the detection mechanism, with the full reference content moved to a follow-up PR targeting the conversation events or install events documentation. Does that work?

Reviewer: Yes, that works.

As it relates to your below comment - let's get enough general information about channelData.app.version that we can _at least_ explain it in a general way here. Like, what event payloads is it on or not on? Is it the right way to get installed app version regardless of what context you're in?

> When a user installs or upgrades your bot, Teams sends an `installationUpdate` activity. After the Sessions feature is enabled, Teams includes the app version in the event payload through `channelData.app.version` for all 1:1 thread scopes, regardless of whether the bot opts into sessions.

Reviewer: I don't follow this, for a couple reasons:

> After the Sessions feature is enabled [...] regardless of whether the bot opts into sessions

I don't think that's right?

This also doesn't say anything about comparing version identifiers and seems to imply that `channelData.app.version` is _not_ included in the event payload if Sessions is not enabled. Is detection about examining the version, or simply seeing if the value is present? If it's the latter, it needs very explicit explanation, because that's very unintuitive.

Also - why would I _want_ to detect session support? (There might be a good answer but I don't know what it is, especially if sessions are almost completely transparent in the agent's implementation).

Author: checking with the engineering team on all three points:

- Is channelData.app.version truly sent regardless of session opt-in, or only when the bot opts in?

- Is detection about the presence of the field, or comparing version strings?

-What's the practical use case for detection if sessions are transparent?

The engineering team provided this content stating it's sent 'for all 1:1 thread scopes regardless of opt-in,' but I agree the explanation needs to be much clearer about the 'why' and 'how' of detection. Will follow up.

@yogapat, could you please help us on this questions?

Create sessions that represent clear goals or workflows. For example, you might create separate sessions for individual tickets, pull requests, or customer requests.

Clear session separation helps users quickly locate and resume conversations.

### Limit creating too many sessions

Create a new session only when it represents a distinct task or workflow that benefits from its own conversation context. Avoid creating sessions for transient updates or single-message interactions. Too many sessions make it harder for users to find relevant conversations in the sessions panel.

### Use a dedicated session for notifications

If your bot sends updates to users based on activity outside of Teams and unrelated to chat history, such as CI/CD results, service alerts, or external system notifications, route them to a single dedicated notifications session instead of creating a new session for each update. Creating a session per notification floods the sessions panel and makes it difficult for users to find task-specific conversations. Store the conversation ID of the notifications session after you create it, and reuse that ID for subsequent messages.

### Provide meaningful first messages

> The first message becomes the session title. Use clear and descriptive messages so users understand the purpose of the session.

Reviewer: Should they be short also?
